### PR TITLE
HostName Handler

### PR DESCRIPTION
**Pull request by Raven0us**
Added the possibility of detecting language based on host name. A lot of sites use alternative URLs because the content is completely different based on the host. 

See `README.md`.

### CODE REVIEW

**Comment by bididev** [comment]
Travis fails on PHP 5.4 due to `::class`.

Yii2 is getting rid of `::className()`, popular extensions such as https://github.com/yiisoft/yii2-queue - https://github.com/yiisoft/yii2-queue/blob/master/src/Queue.php are using `::class`. I can fix that, but do people still use PHP 5.4?

**Comment by cetver** [comment]
Hello, thank you for your interest.

> Yii2 is getting rid of ::className(), popular extensions such as https://github.com/yiisoft/yii2-queue - https://github.com/yiisoft/yii2-queue/blob/master/src/Queue.php are using ::class. I can fix that, but do people still use PHP 5.4?

Yii2 supports PHP starting 5.4 https://github.com/yiisoft/yii2/blob/master/composer.json#L71

**Comment by bididev** [comment]
Yes, okay, I fixed that.

File: handlers/HostNameHandler.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * @var string|Request the Request component ID.
+     */
+    public $request = 'request';
+
+    /** @var array
+     *
+     * key represents the host name and the value represents the language code
+     */
+    public $hostMap = [];
```
Comment (cetver) [inline handlers/HostNameHandler.php]: Please add the possibility to create the host mapping "on fly"
```
    /**
     * @var array|callable Mapping between the hostname and the language or callable function that returns it.
     */
    public $hostMap;
```

File: handlers/HostNameHandler.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * @var array
+     */
+    private $_languages = [];
```
Comment (cetver) [inline handlers/HostNameHandler.php]: This is unnecessary property

File: handlers/HostNameHandler.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            throw new InvalidConfigException(sprintf(
+                'The component with the specified ID "%s" must be an instance of "%s"',
+                $this->request,
+                Request::className()
+            ));
+        }
+
+        $this->request = $request;
+
+        if (!is_array($this->hostMap)) throw new InvalidConfigException("hostMap must be an array");
```
Comment (cetver) [inline handlers/HostNameHandler.php]: Transform into
```
if (is_callable($this->hostMap)) {
    $this->hostMap = call_user_func($this->hostMap);
}

if (!is_array($this->hostMap)) {
    throw new InvalidConfigException(
        'The "hostMap" property must be an array or callable function that returns an array'
    );
}
```

File: handlers/HostNameHandler.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getLanguages()
+    {
```
Comment (cetver) [inline handlers/HostNameHandler.php:57]: too overloaded
```
return (isset($this->hostMap[$this->request->hostName])) ? [$this->hostMap[$this->request->hostName]] : [];
```

File: handlers/HostNameHandler.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace cetver\LanguagesDispatcher\handlers;
+
+use Yii;
+use yii\base\InvalidConfigException;
+use yii\helpers\ArrayHelper;
+use yii\web\Request;
+
+/**
+ * Class HostNameHandler
```
Comment (cetver) [inline handlers/HostNameHandler.php]: Make it more informative:

Class HostNameHandler handles the languages based on the host map.

File: tests/unit/handlers/HostNameHandlerTest.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+        $this->tester->expectException($invalidConfigExceptionClassName, function () {
+            $this->mockWebApplication();
+            new HostNameHandler([
+                'hostMap' => 'whatever'
+            ]);
+        });
+
+        $handler = new HostNameHandler();
+        $this->tester->assertInstanceOf(Request::className(), $handler->request);
```
Comment (cetver) [inline tests/unit/handlers/HostNameHandlerTest.php]: Extend the test, based on https://github.com/cetver/yii2-languages-dispatcher/pull/2/files#r197939078

File: handlers/HostNameHandler.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * @package cetver\LanguagesDispatcher\handlers
+ *
+ * @property array          $hostMap
```
Comment (cetver) [inline handlers/HostNameHandler.php]: This is and 2 lines below are unnecessary, because properties are defined

**Review by cetver — COMMENTED** [review]
Please resolve the review notes

File: tests/unit/handlers/HostNameHandlerTest.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public function testGetLanguages()
+    {
+        $this->mockWebApplication();
```
Comment (cetver) [inline tests/unit/handlers/HostNameHandlerTest.php]: add the test case, based on https://github.com/cetver/yii2-languages-dispatcher/pull/2/files#r197939078

**Pull request merged**